### PR TITLE
The Local LAN API is now used through pdf/golifx

### PR DESCRIPTION
**Pull request by pmaene**
I have refactored the source to use LFIX' Local LAN API, relying on the pdf/golifx implementation (as suggested in the comments of #1). This is the first time I've used Go, so any comments for improvements are welcome!

### CODE REVIEW

**Comment by brutella** [comment]
Looks pretty good. 
- I see cases where calling `light.SetPower(false)` results in two `EventUpdatePower` events – the first with `Power` is `false` and the second one is `true`. The LIFX bulb actually sets the power to false, but the second event results in setting the wrong power state in the bridge. This results in an incorrect behaviour when a user sets the light power switch to off, he sees the light goes down but the app shows power on. Is it just me or do you see the same behaviour?
- the definition of `client` (hklifx.go:38) is not needed anymore
- reduce the log outputs by either removing the log statements or by using `github.com/brutella/log` and disable the info log level with `log.Info = false`

**Comment by pmaene** [comment]
I am also seeing the same behavior with my bulbs. I'll look into its cause, and make the changes to address your other two comments!

I am also having an issue with multiple bulbs: when I try to pair the second bulb in the app, it freezes after selecting it, and the pin dialog is never shown. Do you have any idea whether this is caused by my implementation, or a more general problem with HomeControl?

**Comment by brutella** [comment]
> I am also having an issue with multiple bulbs: when I try to pair the second bulb in the app, it freezes after selecting it, and the pin dialog is never shown. Do you have any idea whether this is caused by my implementation, or a more general problem with HomeControl?

I'm currently looking into this but this seems to be an iOS 9 or OS X issue because it also appears when using the HomeKit Accessory Simulator from the Developer Tools.

**Comment by brutella** [comment]
You should definitely checkout the latest changes of HomeControl, which fixes connection problems on iOS 9.

**Comment by brutella** [comment]
The issue that the power state is reseted to `true` should be fixed with this PR https://github.com/pdf/golifx/pull/1

Now waiting for a merge...

**Comment by pmaene** [comment]
Wow, that's great :). Thanks!

**Comment by brutella** [comment]
So @pdf fixed the power state issue in `v0.3.0`

**Comment by pmaene** [comment]
Yeah, I followed your conversation :). I am currently abroad for work, but I will fix the PR on Sunday!

**Comment by brutella** [comment]
I've just added min and max values for hue, saturation and brightness in commit [a9e62dc](https://github.com/brutella/hc/commit/a9e62dce53bbda67574c87217c1ac5015aa1af43) – if you rebase on the master you can use them to transform between HomeKit and LIFX color values

**Comment by brutella** [comment]
In commit [a9e62dc](https://github.com/brutella/hc/commit/a9e62dce53bbda67574c87217c1ac5015aa1af43) I've added min and max values for hue, saturation and brightness

**Comment by pmaene** [comment]
All right, I've modified the code to use the cached values on an update event. I'm still having problems adding a second bulb, though. The server seems to get stuck in the handshake.

**Comment by pmaene** [comment]
The cached values should probably also be used in a smarter way, as we're now also ignoring any events caused by any other clients, resulting in an inconsistent state in HC.

**Comment by wdvr** [comment]
Which iOS client do you use? I tested your new code with Insteon+ and am able to add multiple 2.0 bulbs, and manage them both individually and together.

Thanks for the update!

**Comment by pdf** [comment]
> The cached values should probably also be used in a smarter way, as we're now also ignoring any events caused by any other clients, resulting in an inconsistent state in HC.

If change events arrive via broadcast, golifx will process them regardless of whether it's the source, but most clients will switch to unicast.  You could poll on some interval, but with many devices that could be heavy.  You will at least synchronize status every 30 seconds when the discovery interval ticks, and appropriate events will be pushed out.

I don't know what communication looks like in HK, but it may make sense to explicitly request state for certain individual groups or devices on the back of some requests, rather than relying on cached state.

**Comment by brutella** [comment]
@pdf What do you mean by "most clients will switch to unicast"? I thought that every state change of the device will be broadcasted to other connected clients?

**Comment by pdf** [comment]
@brutella nope, in fact devices don't send anything on change requests unless they're asked to, and if a response is requested, replies are pretty much exclusively sent back to the originating IP/port via unicast, even when a client makes a broadcast request.  Generally clients will also be making only unicast requests to individual devices unless they're changing global state, after discovery.

**Comment by brutella** [comment]
Every time the bridge gets a request for a characteristic value (e.g. power state), we could fetch the state directly from the devices. This would ensure that we don't return old state. Unfortunately this results in a lot of unnecessary traffic when multiple clients are connected.

I think we should test the current implementation before optimising prematurely. I'm currently fine to get a state synchronisation every 30 second along side reachability updates.

**Comment by pdf** [comment]
Yeah, see how you go.  If you find it's not responsive enough when other clients are involved, one option to keep throughput under control that should be pretty easy to implement would be adding local batching or rate limiting for requests.  Have fun!

**Comment by brutella** [comment]
@pdf Thanks for your awesome library

**Comment by pmaene** [comment]
Yes, absolutely! And for all your tips on how to use it.

File: hklifxd.go [hunk 1]
```diff
@@ -178,39 +216,45 @@ func lightForBulb(bulb *lifx.Bulb) *lifxLight {
...
+	pinArg := flag.String("pin", "", "PIN used to pair the LIFX bulbs with HomeKit")
+	verboseArg := flag.Bool("v", false, "Whether or not log output is displayed")
 
 	flag.Parse()
+
 	pin = *pinArg
 
-	lights = map[string]*lifxLight{}
+	if !*verboseArg {
+		log.Info = false
```
Comment (brutella) [inline hklifxd.go:247]: We also have to set `Verbose` to `false`

**Comment by brutella** [comment]
I just experienced a crash while running the bridge for a longer period of time.
Could you please look into that?

```
2015/10/24 12:03:12 [INFO] Expired Location
2015/10/24 12:03:14 [INFO] Expired Group
2015/10/24 12:03:14 [INFO] Expired Device Stehlampe
panic: interface conversion: interface is common.EventExpiredDevice, not common.EventNewDevice
```

Is there anything you want to add to this pull request? 
Otherwise I would like to merge it in the near future.

**Comment by pmaene** [comment]
I am using the cached values to work around the glitchy power state we encountered initially. I have fixed the bug; it was a silly copy-paste error.

**Comment by pmaene** [comment]
I think this PR is pretty much done... Unless there's something you think is missing?

**Comment by pdf** [comment]
@Pmaene FYI, the cached values are used internally in golifx where appropriate (in groups, so we don't generate recursive requests).  You can certainly use them any time you want to query a device directly without generating a network request, but on the back of events there's no point asking the device for the state, because it's already included in the event.

**Comment by brutella** [comment]
@pdf I thought the same – using the power state from the event is the same.
@Pmaene could you please revert the changes from 20a90ad9eb840aad683b42efe8dcd600a84dad0a

**Comment by pmaene** [comment]
Done :). Looking back, it didn't make sense to do that at all.

**Comment by pmaene** [comment]
Thanks a lot for your help, @pdf and @brutella!

**Comment by brutella** [comment]
Great, please run `go fmt` and then I merge it

**Comment by brutella** [comment]
Good work 👍

**Pull request merged**